compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java

```diff
@@ -175,45 +175,86 @@ public BallerinaSemanticModel(BLangPackage bLangPackage, CompilerContext context
+    }
 
-        BLangNode node = new NodeFinder(false)
-                .lookupEnclosingContainer(this.bLangPackage, symbolLocation.get().lineRange());
+    @Override
+    public List<Location> references(Document sourceDocument, LinePosition position, boolean withDefinition) {
+        return getReferences(sourceDocument, null, position, withDefinition);
+    }
 
-        ReferenceFinder refFinder = new ReferenceFinder(withDefinition);
-        return refFinder.findReferences(node, getInternalSymbol(symbol));
+    @Override
+    public List<Location> references(Symbol symbol, Document targetDocument, boolean withDefinition) {
+        return getReferences(symbol, targetDocument, withDefinition);
     }
 
     @Override
-    public List<Location> references(Document sourceDocument, LinePosition position, boolean withDefinition) {
-        BLangCompilationUnit compilationUnit = getCompilationUnit(sourceDocument);
-        SymbolFinder symbolFinder = new SymbolFinder();
-        BSymbol symbolAtCursor = symbolFinder.lookup(compilationUnit, position);
+    public List<Location> references(Document sourceDocument,
+                                     Document targetDocument,
+                                     LinePosition position,
+                                     boolean withDefinition) {
+        return getReferences(sourceDocument, targetDocument, position, withDefinition);
+    }
+
+    private List<Location> getReferences(Symbol symbol, Document targetDocument, boolean withDefinition) {
+        return getReferences(null, targetDocument, symbol, null, withDefinition);
+    }
+
+    private List<Location> getReferences(Document sourceDocument,
+                                         Document targetDocument,
+                                         LinePosition position,
+                                         boolean withDefinition) {
+        return getReferences(sourceDocument, targetDocument, null, position, withDefinition);
+    }
+
+    private List<Location> getReferences(Document sourceDocument,
+                                         Document targetDocument,
+                                         Symbol symbol,
+                                         LinePosition position,
+                                         boolean withDefinition) {
+
+        Optional<Location> symbolLocation;
+        BSymbol symbolAtCursor;
+
+        if (symbol != null) {
+            symbolLocation = symbol.getLocation();
+            symbolAtCursor = getInternalSymbol(symbol);
+        } else {
+            BLangCompilationUnit sourceCompilationUnit = getCompilationUnit(sourceDocument);
+            SymbolFinder symbolFinder = new SymbolFinder();
+            symbolAtCursor = symbolFinder.lookup(sourceCompilationUnit, position);
+            if (symbolAtCursor == null) {
+                return Collections.emptyList();
+            }
+            symbolLocation = Optional.of(symbolAtCursor.getPosition());
+        }
```

Comment:
Is this the best way to handle this? 🤔  Feel like this method is trying to do too many things. Sometimes an indicator of this is having if-else statements like this one, where we do different different things based on some condition. Instead of doing that, shall we try to do the following:
- Identify the core, high level flow of steps this method is supposed to do to find the references
- Identify the input required for those steps. 
- Identify which of these input need to be passed down to this method. 
- Make those inputs params of the method and move the logic for deriving those input to the relevant overloaded version which calls this method.
e.g., In this if case, you are handling the 2 cases where in one you have the symbol and the other where you have to look it up. Instead of doing that, maybe pass the symbol as a param?

Reply:
Aren't we introduce unnecessary methods with this approach?